File: Tests/ConnectorControlCoreTests/CollectionsFileTests.swift
import XCTest
import ConnectorControlTestSupport
@testable import ConnectorControlCore

/// Mirror: windows/tests/ConnectorControl.Core.Tests/CollectionsFileTests.cs
final class CollectionsFileTests: XCTestCase {
    var tempDir: TempDir!
    override func setUpWithError() throws { tempDir = TempDir(prefix: "collections") }
    override func tearDownWithError() throws { tempDir.dispose() }

    static let sample = CollectionsFile(collections: [
        "Personal": .init(kind: .local, fileName: nil, relativeToStore: nil, origin: nil, needs: [:], publish: nil,
                          provenance: ["dbt": .init(from: "Data team", author: "Acme Data Platform", date: "2026-09-21")]),
        "Data team": .init(kind: .synced, fileName: "data-team.json", relativeToStore: "../mcp/data-team.json", origin: "6f1c4a2e-1b8d-4b0e-9f0a-3c2d7e8a91e2",
                           needs: ["dbt": ["DBT_TOKEN": .init(hint: "cloud.getdbt.com ▸ Account ▸ API tokens", pointer: JSONPointer(["env", "DBT_TOKEN"])),
                                           "server_path": .init(hint: "your ledger clone, then dist/index.js", pointer: JSONPointer(["args", "1"]))]],
                           publish: nil, provenance: [:]),
        "Consulting": .init(kind: .local, fileName: nil, relativeToStore: nil, origin: nil, needs: [:],
                            publish: .init(slug: "consulting", origin: "0c9b7d1e-5a3f-4f2c-8e6d-2b1a9c8d7e6f",
                                           intent: PublishIntent(shareValues: ["tax": ["LOG_LEVEL"]],
                                                                 pathMarks: ["ledger": [JSONPointer(["args", "1"]): .init(name: "server_path", hint: "your ledger clone, then dist/index.js",
                                                                                                                          value: "/Users/you/ledger/dist/index.js")]],
                                                                 hints: ["tax": ["ANTHROPIC_API_KEY": "console.anthropic.com ▸ API keys"]])),
                            provenance: [:]),
    ])

    func testEncodeDecodeRoundTripAndGolden() throws {
        XCTAssertEqual(try CollectionsFile.decode(Self.sample.encode()), Self.sample)
        // Written once from this Mac's Foundation (CONNECTOR_CONTROL_UPDATE_GOLDENS=1), then compared
        // byte for byte here and by the C# mirror, so both platforms write the same sidecar.
        let url = Fixtures.url("collections.json")
        let data = try Self.sample.encode().serialized()
        if ProcessInfo.processInfo.environment["CONNECTOR_CONTROL_UPDATE_GOLDENS"] == "1" { try data.write(to: url) }
        XCTAssertEqual(data, try Data(contentsOf: url))
        XCTAssertEqual(try CollectionsFile.decode(JSONValue.parse(Data(contentsOf: url))), Self.sample)
    }

    func testKindDefaultsToLocal() {
        XCTAssertEqual(Self.sample.kind(of: "Data team"), .synced)
        XCTAssertEqual(Self.sample.kind(of: "Personal"), .local)
        XCTAssertEqual(Self.sample.kind(of: "Never heard of it"), .local)
    }

    func testLoadTreatsMissingAndCorruptAsEmpty() throws {
        let url = tempDir.file("collections.json")
        XCTAssertEqual(CollectionsFile.load(from: url), CollectionsFile(collections: [:]))
        try Data("{not json".utf8).write(to: url)
        XCTAssertEqual(CollectionsFile.load(from: url), CollectionsFile(collections: [:]))
        XCTAssertTrue(FileManager.default.fileExists(atPath: url.path), "load never moves a file aside")
    }

    func testLoadIfReadableTellsAnUnreadableFileFromAnEmptyOne() throws {
        let url = tempDir.file("collections.json")
        XCTAssertEqual(CollectionsFile.loadIfReadable(from: url), CollectionsFile(collections: [:]), "missing is empty")
        try CollectionsFile(collections: [:]).save(to: url, staging: nil)
        XCTAssertEqual(CollectionsFile.loadIfReadable(from: url), CollectionsFile(collections: [:]), "no entries is empty")
        try Data("{not json".utf8).write(to: url)
        XCTAssertNil(CollectionsFile.loadIfReadable(from: url), "a file that cannot be parsed is unreadable")
        try Data("{\"version\":2,\"collections\":{}}".utf8).write(to: url)
        XCTAssertNil(CollectionsFile.loadIfReadable(from: url), "a file that cannot be decoded is unreadable")
        try Data().write(to: url)
        XCTAssertNil(CollectionsFile.loadIfReadable(from: url), "a file truncated to nothing is unreadable")
        XCTAssertTrue(FileManager.default.fileExists(atPath: url.path), "load never moves a file aside")
    }

    func testSaveIsOwnerOnlyAndReloads() throws {
        let url = tempDir.file("collections.json")
        try Self.sample.save(to: url, staging: nil)
        let mode = try XCTUnwrap(FileManager.default.attributesOfItem(atPath: url.path)[.posixPermissions] as? Int)
        XCTAssertEqual(mode, 0o600)
        XCTAssertEqual(CollectionsFile.load(from: url), Self.sample)
    }

    func testReconcileDropsNamesTheStoreNoLongerHas() {
        let store = MasterStore(activeCollection: "Personal", collections: ["Personal": Collection(), "Consulting": Collection()])
        let reconciled = Self.sample.reconciled(with: store)
        XCTAssertEqual(Set(reconciled.collections.keys), ["Personal", "Consulting"])
    }

    func testAnUnknownVersionDecodesAsMalformed() {
        XCTAssertThrowsError(try CollectionsFile.decode(.object(["version": .int(9), "collections": .object([:])])))
    }

    /// No release wrote a path mark without the value it was made on, but a build from before
    /// values were kept may have: such a mark loads as the pointer-only mark it was.
    func testAPathMarkWithNoValueStillLoads() throws {
        let json = try JSONValue.parse(Data("""
            {"version": 1, "collections": {"Consulting": {"kind": "local", "publish": {
              "slug": "consulting", "origin": "o", "shareValues": {}, "hints": {},
              "paths": {"ledger": {"/args/1": {"name": "server_path", "hint": null}}}}}}}
            """.utf8))
        let marks = try CollectionsFile.decode(json).collections["Consulting"]?.publish?.intent.pathMarks["ledger"]
        XCTAssertEqual(marks, [JSONPointer(["args", "1"]): .init(name: "server_path", hint: nil, value: nil)])
    }

    /// A local collection with nothing to say is absent from the file, not written as an empty
    /// entry: the sidecar only ever records what the master list cannot.
    func testALocalCollectionWithNothingToSayIsNotWritten() throws {
        let file = CollectionsFile(collections: ["Personal": .local, "Consulting": Self.sample.collections["Consulting"]!])
        XCTAssertEqual(file.encode().value(at: JSONPointer(["collections", "Personal"])), nil)
        XCTAssertEqual(try CollectionsFile.decode(file.encode()).collections.keys.sorted(), ["Consulting"])
        XCTAssertEqual(file.kind(of: "Personal"), .local)
    }
}

File: windows/tests/ConnectorControl.Core.Tests/CollectionsFileTests.cs
using System.Runtime.Versioning;
using ConnectorControl.Core;
using ConnectorControl.Core.Tests.TestSupport;

namespace ConnectorControl.Core.Tests;

/// <summary>Mirror: Tests/ConnectorControlCoreTests/CollectionsFileTests.swift</summary>
public sealed class CollectionsFileTests : IDisposable
{
    private readonly TempDir dir = new("collections");

    public void Dispose() => dir.Dispose();

    private static CollectionsFile Sample => new(new Dictionary<string, CollectionsFile.Entry>
    {
        ["Personal"] = new(
            CollectionKind.Local, null, null, null, null, null,
            new Dictionary<string, CollectionsFile.Provenance>
            {
                ["dbt"] = new("Data team", "Acme Data Platform", "2026-09-21"),
            }),
        ["Data team"] = new(
            CollectionKind.Synced, "data-team.json", "../mcp/data-team.json", "6f1c4a2e-1b8d-4b0e-9f0a-3c2d7e8a91e2",
            new Dictionary<string, IReadOnlyDictionary<string, CollectionsFile.Need>>
            {
                ["dbt"] = new Dictionary<string, CollectionsFile.Need>
                {
                    ["DBT_TOKEN"] = new("cloud.getdbt.com ▸ Account ▸ API tokens", new JsonPointer(["env", "DBT_TOKEN"])),
                    ["server_path"] = new("your ledger clone, then dist/index.js", new JsonPointer(["args", "1"])),
                },
            },
            null,
            null),
        ["Consulting"] = new(
            CollectionKind.Local, null, null, null, null,
            new CollectionsFile.PublishRecord("consulting", "0c9b7d1e-5a3f-4f2c-8e6d-2b1a9c8d7e6f", new PublishIntent(
                new Dictionary<string, IReadOnlySet<string>> { ["tax"] = new HashSet<string>(StringComparer.Ordinal) { "LOG_LEVEL" } },
                new Dictionary<string, IReadOnlyDictionary<JsonPointer, PublishIntent.PathMark>>
                {
                    ["ledger"] = new Dictionary<JsonPointer, PublishIntent.PathMark>
                    {
                        [new JsonPointer(["args", "1"])] = new("server_path", "your ledger clone, then dist/index.js",
                            "/Users/you/ledger/dist/index.js"),
                    },
                },
                new Dictionary<string, IReadOnlyDictionary<string, string>>
                {
                    ["tax"] = new Dictionary<string, string>(StringComparer.Ordinal) { ["ANTHROPIC_API_KEY"] = "console.anthropic.com ▸ API keys" },
                })),
            null),
    });

    [Fact]
    public void EncodeDecodeRoundTripAndGolden()
    {
        Assert.Equal(Sample, CollectionsFile.Decode(Sample.Encode()));
        // Written once from the Mac's Foundation, then compared byte for byte there and here,
        // so both platforms write the same sidecar. This side never regenerates it.
        var expected = File.ReadAllBytes(Fixtures.Path("collections.json"));
        Assert.Equal(expected, Sample.Encode().Serialize());
        Assert.Equal(Sample, CollectionsFile.Decode(JsonValue.Parse(expected)));
    }

    [Fact]
    public void KindDefaultsToLocal()
    {
        Assert.Equal(CollectionKind.Synced, Sample.KindOf("Data team"));
        Assert.Equal(CollectionKind.Local, Sample.KindOf("Personal"));
        Assert.Equal(CollectionKind.Local, Sample.KindOf("Never heard of it"));
    }

    [Fact]
    public void LoadTreatsMissingAndCorruptAsEmpty()
    {
        var path = dir.File("collections.json");
        Assert.Equal(new CollectionsFile([]), CollectionsFile.Load(path));
        File.WriteAllText(path, "{not json");
        Assert.Equal(new CollectionsFile([]), CollectionsFile.Load(path));
        Assert.True(File.Exists(path), "load never moves a file aside");
    }

    [Fact]
    public void LoadIfReadableTellsAnUnreadableFileFromAnEmptyOne()
    {
        var path = dir.File("collections.json");
        Assert.Equal(new CollectionsFile([]), CollectionsFile.LoadIfReadable(path));   // missing is empty
        new CollectionsFile([]).Save(path);
        Assert.Equal(new CollectionsFile([]), CollectionsFile.LoadIfReadable(path));   // no entries is empty
        File.WriteAllText(path, "{not json");
        Assert.Null(CollectionsFile.LoadIfReadable(path));   // a file that cannot be parsed is unreadable
        File.WriteAllText(path, "{\"version\":2,\"collections\":{}}");
        Assert.Null(CollectionsFile.LoadIfReadable(path));   // a file that cannot be decoded is unreadable
        File.WriteAllBytes(path, []);
        Assert.Null(CollectionsFile.LoadIfReadable(path));   // a file truncated to nothing is unreadable
        Assert.True(File.Exists(path), "load never moves a file aside");
    }

    [Fact]
    [SupportedOSPlatform("windows")]
    public void SaveIsOwnerOnlyAndReloads()
    {
        var path = dir.File("collections.json");
        Assert.True(Sample.Save(path).Protected);
        Assert.Equal(Sample, CollectionsFile.Load(path));
        Assert.SkipUnless(OperatingSystem.IsWindows(), "Windows only");
        Assert.True(OwnerOnlyAcl.IsOwnerOnly(path));
    }

    [Fact]
    public void ReconcileDropsNamesTheStoreNoLongerHas()
    {
        var store = new MasterStore(2, "Personal", new Dictionary<string, Collection>
        {
            ["Personal"] = new(),
            ["Consulting"] = new(),
        });
        var reconciled = Sample.Reconciled(store);
        Assert.Equal(["Consulting", "Personal"], reconciled.Collections.Keys.Order(StringComparer.Ordinal).ToArray());
    }

    [Fact]
    public void AnUnknownVersionDecodesAsMalformed()
    {
        Assert.Throws<CollectionsFileException>(() => CollectionsFile.Decode(
            JsonValue.Object(("version", JsonValue.Int(9)), ("collections", JsonValue.Object()))));
    }

    /// <summary>
    /// No release wrote a path mark without the value it was made on, but a build from before
    /// values were kept may have: such a mark loads as the pointer-only mark it was.
    /// </summary>
    [Fact]
    public void APathMarkWithNoValueStillLoads()
    {
        var json = JsonValue.Parse("""
            {"version": 1, "collections": {"Consulting": {"kind": "local", "publish": {
              "slug": "consulting", "origin": "o", "shareValues": {}, "hints": {},
              "paths": {"ledger": {"/args/1": {"name": "server_path", "hint": null}}}}}}}
            """);
        var marks = CollectionsFile.Decode(json).Collections["Consulting"].Publish!.Intent.PathMarks["ledger"];
        Assert.Equal(new PublishIntent.PathMark("server_path", null, null), Assert.Single(marks).Value);
        Assert.Equal(new JsonPointer(["args", "1"]), Assert.Single(marks).Key);
    }

    /// <summary>A local collection with nothing to say is absent from the file, not written as an empty entry: the sidecar only ever records what the master list cannot.</summary>
    [Fact]
    public void ALocalCollectionWithNothingToSayIsNotWritten()
    {
        var file = new CollectionsFile(new Dictionary<string, CollectionsFile.Entry>
        {
            ["Personal"] = CollectionsFile.Entry.Local,
            ["Consulting"] = Sample.Collections["Consulting"],
        });
        Assert.Null(file.Encode().ValueAt(new JsonPointer(["collections", "Personal"])));
        Assert.Equal(["Consulting"], CollectionsFile.Decode(file.Encode()).Collections.Keys.Order(StringComparer.Ordinal).ToArray());
        Assert.Equal(CollectionKind.Local, file.KindOf("Personal"));
    }
}
